File: lib/firestore_refs.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import 'visitor_log/visitor_log.dart';

final _db = FirebaseFirestore.instance;

/// visitorLogs コレクションの参照。
final visitorLogsRef = _db.collection('visitorLogs').withConverter(
  fromFirestore: (ds, _) {
    return VisitorLog.fromDocumentSnapshot(ds);
  },
  toFirestore: (obj, _) {
    final json = obj.toJson();
    return json;
  },
);

/// visitorLog ドキュメントの参照。
DocumentReference<VisitorLog> visitorLogRef({required String visitorLogId}) =>
    visitorLogsRef.doc(visitorLogId);

File: functions/src/firestore-refs/firestoreRefs.ts
import { CollectionReference, DocumentReference } from '@google-cloud/firestore'
import * as admin from 'firebase-admin'
import { VisitorLog } from '../models/visitorLog'
import { visitorLogConverter } from './converters'

/** undefined なプロパティを無視するよう設定した db オブジェクト。 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/** visitorLogs コレクションの参照 */
export const visitorLogsRef: CollectionReference<VisitorLog> = db
    .collection(`visitorLogs`)
    .withConverter<VisitorLog>(visitorLogConverter)

/** visitorLog ドキュメントの参照 */
export const visitorLogRef = ({
    visitorLogId
}: {
    visitorLogId: string
}): DocumentReference<VisitorLog> => visitorLogsRef.doc(visitorLogId)
